continue prost

jokemanfire opened the pull request:
work continue with https://github.com/containerd/ttrpc-rust/pull/173

justxuewei committed:
codegen: Introduce a new codegen using prost

This commit refactors the ttrpc-codegen and the compiler, and merges the two
crates into a single crate, named "codegen". The codegen uses prost crate, a
protobuf compiler for Rust.

Signed-off-by: Xuewei Niu <niuxuewei.nxw@antgroup.com>

justxuewei committed:
ttrpc-rust: Support the new version of codegen

The ttrpc provides a "prost" feature to support the new version of
codegen. An "example2" has been added to demonstrate how to use the
codegen.

Signed-off-by: Xuewei Niu <niuxuewei.nxw@antgroup.com>

justxuewei committed:
ci: Add codegen-related ci testing

Install protoc when executing `make deps` of the ttrpc. Add codegen's
check and build, and example2' build to the ci testing.

Signed-off-by: Xuewei Niu <niuxuewei.nxw@antgroup.com>

justxuewei committed:
codegen: Fix issues reported by cargo clippy

Fix all issues reported by cargo clippy to make ci testing pass.

Signed-off-by: Xuewei Niu <niuxuewei.nxw@antgroup.com>

jokemanfire committed:
fix: prost compile faild.

merge from remote branch.
ref: https://github.com/containerd/ttrpc-rust/pull/173

Co-authored-by: Xuewei Niu <niuxuewei.nxw@antgroup.com>
Signed-off-by: jokemanfire <hu.dingyang@zte.com.cn>

Tim-Zhang commented:
@jokemanfire How about reopening https://github.com/containerd/ttrpc-rust/pull/173?

jokemanfire commented:
> @jokemanfire I appreciate you taking on this work very much. Why don't we directly reopen #173 and continue the work? Just let me know if there are any difficulties.
> 
> Thanks agian.

There are some conflicts and API changes that need to be resolved and some code needs to be packaged to adapt to the current API. It's not easy to operate with the previous version of PR. I need to add some commit to complete this feature. I have communicated with the original author.

jokemanfire commented:
@Tim-Zhang  @justxuewei   Local test all pass but i don't know how to install protoc in CI.

Tim-Zhang commented:
@jokemanfire FYI 
https://github.com/marketplace/actions/setup-protoc 
https://www.andreamedda.com/posts/go-buf-github-actions/

jokemanfire committed:
ci: change protoc install script in ci

fix ci for protoc

Signed-off-by: jokemanfire <hu.dingyang@zte.com.cn>

jokemanfire committed:
fix: fix test in protoc async

because of the api change
1. fix the marco in prost
2. fix the test

Signed-off-by: jokemanfire <hu.dingyang@zte.com.cn>

jokemanfire commented:
Basically enough @Tim-Zhang  @justxuewei  just take a loop . :)

justxuewei commented:
Nice work, thanks @jokemanfire! Could you cleanup your commits? It makes us easier to review and maintain.

1. Use `git rebase` instead of `git merge`.
2. Modify the commit title to `{subsystem}: {brief title for your changes}`.
3. Split your commits, and don't mix bugfixes, refactoring and new features.

Thanks!

jokemanfire committed:
fix: fix the rebase caused the error

fix rebase error.

Signed-off-by: jokemanfire <hu.dingyang@zte.com.cn>

jokemanfire commented:
> Nice work, thanks @jokemanfire! Could you cleanup your commits? It makes us easier to review and maintain.好的工作，谢谢！你能清理一下你的提交吗？这让我们更容易审查和维护。
> 
> 1. Use `git rebase` instead of `git merge`.使用 `git rebase` 代替 `git merge` 。
> 2. Modify the commit title to `{subsystem}: {brief title for your changes}`, and give it a brief introduction in the commit message.修改提交标题为 `{subsystem}: {brief title for your changes}` ，并在提交信息中简要介绍。
> 3. Split your commits, and don't mix bugfixes, refactoring and new features.拆分您的提交，不要混合修复错误、重构和新功能。
> 4. Please modify this PR title to something like "codegen: Introduce a new codegen using prost".请将此 PR 标题修改为类似于“codegen：引入使用 prost 的新代码生成器”的标题。
> 
> Thanks!  谢谢！

I have changed it.

jokemanfire committed:
fix(example2): skip build example2 in windows

skip build example2 in windows

Signed-off-by: jokemanfire <hu.dingyang@zte.com.cn>

jokemanfire commented:
> Once this feature gets merged, I prefer to bump a major version to 2 to avoid some compatibility issues, as it nearly changes the entire codegen engine. WDYT? 一旦这个功能合并，我更倾向于将主版本号提升到 2，以避免一些兼容性问题，因为它几乎改变了整个代码生成引擎。你怎么看？@jokemanfire @Tim-Zhang

This modification is indeed quite significant, and I agree. It depends on the maintainer's thoughts

Tim-Zhang commented on build.rs:
```diff
@@ -7,6 +7,11 @@ fn main() {
     let path: PathBuf = [out_dir.clone(), "mod.rs".to_string()].iter().collect();
     fs::write(path, "pub mod ttrpc;").unwrap();
 
+    generate_ttrpc(&out_dir);
```
To get unified internal APIs which was defined in ttrpc.proto, how about keeping build script untouched?

I think we can try to avoid so many condition compiling sentences like  #[cfg(not(feature = "prost"))] in code.

jokemanfire replied:
right ,I think adding a transformation to the camel hump naming convention can reduce this conditional compilation code

jokemanfire committed:
fix(build.rs): reduce the  Code which is enum's difference

while use protoc,rename the Code enum.

Signed-off-by: jokemanfire <hu.dingyang@zte.com.cn>

jokemanfire committed:
fix(macro): change marco's enum to same

fix marco

Signed-off-by: jokemanfire <hu.dingyang@zte.com.cn>

jokemanfire commented:
I tried to minimize conditional compilation as much as possible @Tim-Zhang  @justxuewei

Tim-Zhang reviewed (dismissed):
Thanks for continuing the Prost support work. I found several correctness and CI issues that should be addressed before this can be merged.

### Correctness issues

1. **The documented Prost configuration cannot compile** (`README.md:151`)

   The example enables `prost` while leaving the default `rustprotobuf` feature enabled, so the new mutual-exclusion check aborts compilation. Please document `default-features = false`.

2. **Qualified request types break generated handlers** (`codegen/src/svcgen.rs:158`)

   Imported request types may be paths such as `super::google::protobuf::Empty`. Passing such a path to `format_ident!` panics because it is not a single identifier. Please use path-aware token conversion, as is already done for trait signatures.

3. **Shared imports are generated once per service** (`codegen/src/svcgen.rs:26`)

   prost-build appends every service in a package to the same module. Generating `HashMap`, `Arc`, `Message`, and `async_trait` imports for every service causes E0252 for packages containing multiple services. These imports need to be emitted once per generated module/package.

### Additional issues

4. **docs.rs enables both protobuf backends** (`src/lib.rs:84`)

   The existing `all-features = true` metadata enables both `rustprotobuf` and `prost`, so docs.rs will fail the mutual-exclusion check. Please configure docs.rs to select one backend.

5. **Generated handler names can collide across services** (`codegen/src/svcgen.rs:139`)

   Two services in one package that both define `Get` generate the same `GetMethod` type. Include the service name in handler type names.

6. **The service and serde generation switches are not honored independently** (`codegen/src/codegen.rs:79`)

   The service generator is installed unconditionally, while serde attributes are only configured when `generate_service` is true. Thus `set_generate_service(false).set_serde(true)` still emits service code and omits serde derives.

7. **Package-less services get an invalid leading separator** (`codegen/src/svcgen.rs:227`)

   A proto without a package generates names such as `.Health` and `/.Health/Check`. The package separator should only be added when the package is non-empty.

8. **The Prost sync server bypasses the response-size limit** (`src/sync/utils.rs:47`)

   Oversized Prost responses are sent in full instead of being replaced by the small RPC error used by the rust-protobuf path. Please retain the `MESSAGE_LENGTH_MAX` behavior for both backends.

9. **The Prost sync client checks only payload size** (`src/sync/client.rs:161`)

   Service, method, metadata, and framing can make the encoded `Request` exceed the limit even when its payload does not. Please check `req.encoded_len()` or the encoded buffer length.

10. **The Windows OS variable is not expanded correctly** (`Makefile:26`)

    GNU make interprets `$OS` as `$O` followed by `S`. Please use `$(OS)`.

11. **The Windows multi-command build step should keep Bash** (`.github/workflows/bvt.yml:37`)

    With the default PowerShell shell, an earlier native `make` failure may be hidden if the final command succeeds. Please retain `shell: bash` so intermediate failures terminate the step.

Please also add code-generation fixtures covering imported request types, multiple services with overlapping method names, and package-less schemas.

Tim-Zhang commented on README.md:
```diff
@@ -138,3 +138,19 @@ cargo update
+1. The protoc should be installed.
+2. Enabling "prost" feature for the ttrpc-rust.
+3. The Rust files are named based on their package name, rather than the proto
+   filename, e.g. `ttrpc = { version = "1.0", features = ["prost"] }`.
```
This example enables `prost` while keeping default features, which also enables `rustprotobuf`; the mutual-exclusion check then aborts compilation. Please document `default-features = false` and include the desired runtime feature explicitly, for example `features = ["sync", "prost"]`.

Tim-Zhang commented on src/lib.rs:
```diff
@@ -76,3 +76,17 @@ cfg_async! {
+    () => {};
+    ($first:tt $(,$rest:tt)*) => {
+        $(
+            #[cfg(all(feature = $first, feature = $rest))]
```
`docs.rs` is configured with `all-features = true`, so it activates both protobuf backends and always reaches this compile error. Configure docs.rs with one explicit compatible backend feature set instead of all features.

Tim-Zhang commented on codegen/src/svcgen.rs:
```diff
@@ -0,0 +1,503 @@
+    fn method_handler_impl_sync_token(&self, struct_name: &Ident, method: &Method) -> TokenStream {
+        let mod_path = ttrpc_mod();
+        let context = self.ttrpc_context(false);
+        let input_type = format_ident!("{}", method.input_type);
```
`method.input_type` can be a qualified path such as `super::google::protobuf::Empty`. Passing it to `format_ident!` panics because it is not a single identifier. Use the path-aware `type_token(&method.input_type)` conversion here and in the async handler.

Tim-Zhang commented on codegen/src/svcgen.rs:
```diff
@@ -0,0 +1,503 @@
+    fn finalize_package(&mut self, _package: &str, _buf: &mut String) {}
+    /// Generate services
+    fn generate(&mut self, service: Service, buf: &mut String) {
+        self.generate_type_aliases(buf);
```
`prost-build` calls `generate` once per service and appends services from a package into the same module. Emitting `HashMap`, `Arc`, `Message`, and `async_trait` imports here for every service causes E0252 on packages with multiple services. Emit package-level imports once, for example from `finalize_package`.

Tim-Zhang commented on codegen/src/svcgen.rs:
```diff
@@ -0,0 +1,503 @@
+    }
+
+    fn method_handler_token(&self, service: &Service, method: &Method) -> TokenStream {
+        let struct_name = format_ident!("{}Method", to_camel_case(method.proto_name.as_str()));
```
The handler type is derived only from the method name, so `Foo.Get` and `Bar.Get` in one package both generate `GetMethod`. Include the service name in this type, for example `FooGetMethod`.

Tim-Zhang commented on codegen/src/codegen.rs:
```diff
@@ -0,0 +1,189 @@
+    fn compile_protos(&self) -> Result<()> {
+        let mut config = Config::new();
+        config.out_dir(self.out_dir.as_ref());
+        config.service_generator(Box::new(TtrpcServiceGenerator::new(self.async_mode)));
```
The service generator is installed unconditionally, while serde attributes are nested under `generate_service`. Consequently `set_generate_service(false).set_serde(true)` still emits services and omits serde derives. Configure these two switches independently.

Tim-Zhang commented on codegen/src/svcgen.rs:
```diff
@@ -0,0 +1,503 @@
+        let service_trait = format_ident!("{}", service.name);
+        let mod_path = ttrpc_mod();
+        let method_inserts: Vec<_> = service.methods.iter().map(|method| {
+            let key = format!("/{}.{}/{}", service.package, service.name, method.proto_name);
```
When `service.package` is empty this produces `/.Health/Check`; the async service path similarly becomes `.Health`. Build the package separator conditionally so package-less schemas use `/Health/Check` and `Health`.

Tim-Zhang commented on src/sync/utils.rs:
```diff
@@ -35,6 +38,27 @@ pub fn response_to_channel(
+    tx: std::sync::mpsc::Sender<(MessageHeader, Vec<u8>)>,
+) -> Result<()> {
+    let mut buffer = Vec::new();
+    <Response as prost::Message>::encode(&res, &mut buffer).map_err(err_to_others_err!(e, ""))?;
```
The Prost branch serializes and sends the response without the `MESSAGE_LENGTH_MAX` check used above. This bypasses the established server-side size limit. Encode through the shared codec and preserve the same oversized-response replacement behavior for both backends.

Tim-Zhang commented on src/sync/client.rs:
```diff
@@ -157,6 +158,10 @@ impl Client {
     }
     pub fn request(&self, req: Request) -> Result<Response> {
+        #[cfg(feature = "prost")]
+        check_oversize(req.payload.len(), false)?;
```
This checks only the RPC payload, but service, method, metadata, and protobuf framing can make the complete encoded `Request` exceed the limit. Check `req.encoded_len()` or the final encoded buffer length instead.

Tim-Zhang commented on Makefile:
```diff
@@ -21,12 +23,24 @@ build: debug
 .PHONY: test
 test:
-	cargo test --all-features --verbose
+ifeq ($OS,Windows_NT)
```
GNU make parses `$OS` as the one-character variable `$O` followed by `S`, so this Windows branch is never selected. Use `$(OS)`.

Tim-Zhang commented on .github/workflows/bvt.yml:
```diff
@@ -23,15 +26,16 @@ jobs:
-        # It's important for windows to fail correctly
-        # https://github.com/actions/runner-images/issues/6668 
-        shell: bash
+          make -C example2 build-examples
```
Removing `shell: bash` makes this multi-command step use PowerShell on Windows, where an earlier native-command failure can be hidden by a later successful command. Keep `shell: bash`, or split the commands into separate steps with explicit failure handling.

Tim-Zhang commented on build.rs:
```diff
@@ -20,3 +25,70 @@ fn main() {
+    ];
+
+    // replace the enum value in the file
+    let mut modified_content = content.clone();
```
Refactor/performance: please avoid rewriting Prost-generated Rust as text. This is coupled to Prost's formatting and performs many full-string replacement passes; `content.clone()` also duplicates the whole file immediately. Preserve the public enum spelling through a small compatibility layer/associated constants or generator configuration. If this remains temporarily, at least move `content` instead of cloning it.

Tim-Zhang commented on codegen/src/svcgen.rs:
```diff
@@ -0,0 +1,503 @@
+        };
+        quote!(
+            struct #struct_name {
+                service: Arc<Box<dyn #service_name + Send + Sync>>,
```
Refactor/performance: `Arc<Box<dyn Trait>>` adds an allocation and pointer indirection, and diverges from the existing rust-protobuf generator API, which uses `Arc<dyn Trait + Send + Sync>`. Use `Arc<dyn ...>` in the handler fields and both `create_*` signatures.

Tim-Zhang commented on src/asynchronous/utils.rs:
```diff
@@ -49,8 +50,42 @@ macro_rules! async_request_handler {
 
+/// Handle request in async mode.
+#[macro_export]
+#[cfg(feature = "prost")]
```
Refactor: duplicating each complete handler/client macro for the two backends makes control-flow fixes easy to apply to only one side. Extend the existing `Codec` abstraction and add small backend-specific request/response/status helpers, then keep one sync and one async handler implementation.

Tim-Zhang commented on codegen/src/codegen.rs:
```diff
@@ -0,0 +1,189 @@
+    }
+}
+
+#[derive(Default)]
```
Idiomatic Rust: the boolean and enum settings can store their defaults directly rather than using `Option`, and `build` should normally consume `self`. The `P: Default` bound is unrelated to paths, while borrowing all inputs makes this builder unnecessarily restrictive. Consider owned `PathBuf`/`Vec<PathBuf>` fields and `build(self)`.

Tim-Zhang commented on Cargo.toml:
```diff
@@ -34,11 +35,14 @@ tokio-vsock = { version = "0.7.0", optional = true }
 # lock home to avoid conflict with latest version
 home = "=0.5.9"
 protobuf-codegen = "3.1.0"
+prost-build = { version = "0.13", optional = true }
```
Refactor/dependencies: the runtime uses `prost` 0.11 here while the root build uses `prost-build` 0.13, and the new codegen crate pins the Prost family to 0.11. Keep the Prost crates on one compatible version and centralize the versions to avoid duplicate dependency trees and generated/runtime drift.

Tim-Zhang commented on codegen/src/util.rs:
```diff
@@ -0,0 +1,97 @@
+    let mut camel_case_name = String::with_capacity(name.len());
+    for s in NameSpliter::new(name) {
+        let mut chs = s.chars();
+        camel_case_name.extend(chs.next().unwrap().to_uppercase());
```
Refactor/correctness: this custom name splitter can yield an empty segment after leading underscores and then panic at this `unwrap`; it also reimplements casing already handled by established Rust casing utilities and by normalized `prost-build` names. Prefer those normalized names or a well-tested casing helper.

Tim-Zhang commented on example-prost/Cargo.toml:
```diff
@@ -0,0 +1,55 @@
+[package]
```
Scope/refactor: `example2` duplicates the existing examples and brings in large third-party proto trees solely to switch protobuf backends, so future fixes and tests can diverge. Prefer running the existing example under a backend feature matrix and add small focused fixtures only for Prost-specific codegen behavior.

Tim-Zhang commented on .gitignore:
```diff
@@ -6,4 +6,9 @@ Cargo.lock
+example2/protocols/**/*.rs
+!example2/protocols/**/mod.rs
 src/ttrpc.rs
+example2/protocols/**/*.rs
```
These `example2` ignore/negation rules duplicate lines 10-11. Please remove the redundant entries and keep the file's final newline.

Tim-Zhang reviewed (changes requested):
Several generator edge cases, size-limit regressions, and CI issues remain. The inline comments also include refactoring and idiomatic-Rust feedback focused on containing backend-specific behavior, reducing duplication, and avoiding unnecessary allocations and API divergence. Please address the inline findings before merge and add focused fixtures for imported types, multiple services with overlapping method names, and package-less schemas.

Tim-Zhang commented:
@jokemanfire I have left a few review comments, and btw please resolve the conflicts, thanks.

jokemanfire commented:
> @jokemanfire I have left a few review comments, and btw please resolve the conflicts, thanks.

thanks , I will take some time to resolve it.

jokemanfire commented on src/sync/utils.rs:
```diff
@@ -78,8 +103,41 @@ macro_rules! request_handler {
+    ($class: ident, $ctx: ident, $req: ident, $req_type: ident, $req_fn: ident) => {
+        let mut req = $req_type::default();
+        req.merge(&$req.payload as &[u8])
+            .map_err(::ttrpc::err_to_others!(e, ""))?;
```
please  check  I see the doc not need  mut  https://docs.rs/prost/0.14.4/prost/trait.Message.html#method.merge

jokemanfire commented on src/sync/client.rs:
```diff
@@ -158,6 +159,10 @@ impl Client {
+        check_oversize(req.payload.len(), false)?;
+
+        #[cfg(not(feature = "prost"))]
         check_oversize(req.compute_size() as usize, false)?;
```
I think this shoud not be modified

jokemanfire commented on src/sync/server.rs:
```diff
@@ -162,28 +166,72 @@ fn start_method_handler_thread(
+
+            #[cfg(feature = "prost")]
+            {
+                if let Err(x) = req.merge(&buf as &[u8]) {
```
incorrect

jokemanfire commented on src/asynchronous/utils.rs:
```diff
@@ -49,8 +50,42 @@ macro_rules! async_request_handler {
+    ($class: ident, $ctx: ident, $req: ident, $req_type: ident, $req_fn: ident) => {
+        let mut req = $req_type::default();
+        req.merge(&$req.payload as &[u8])
+            .map_err(::ttrpc::err_to_others!(e, "Merge request.payload"))?;
```
incorrect

Tim-Zhang commented on codegen/src/codegen.rs:
```diff
@@ -0,0 +1,185 @@
+    }
+}
+
+pub struct CodegenBuilder {
```
Since this crate keeps the `ttrpc-codegen` package name and replaces the existing implementation, please preserve the current fluent API instead of introducing a separate `CodegenBuilder` API.

The expected usage should remain similar to:

```rust
Codegen::new()
    .out_dir(...)
    .inputs(...)
    .include(...)
    .prost()
    .customize(...)
    .run()
```

Migrating the implementation to Prost should not require users to rewrite their build scripts around `set_out_dir`, `set_protos`, `set_includes`, `build`, and `generate`. It would also be preferable to keep the existing public error abstraction instead of exposing `anyhow::Result` as part of the codegen API.

Tim-Zhang commented on codegen/src/codegen.rs:
```diff
@@ -0,0 +1,185 @@
+    fn compile_protos(&self) -> Result<()> {
+        let mut config = Config::new();
+        config.out_dir(&self.out_dir);
+        config.service_generator(Box::new(TtrpcServiceGenerator::new(self.async_mode)));
```
`service_generator` is installed unconditionally, while the `serde` configuration is nested under `generate_service`.

As a result, `generate_service = false` still generates service code, and `serde = true` is ignored unless service generation is enabled.

The existing `ttrpc-codegen` API always generates service code, so please remove `generate_service` and continue generating services unconditionally. If the option is retained, it needs to be honored independently from the Serde option.

Tim-Zhang commented on codegen/src/svcgen.rs:
```diff
@@ -0,0 +1,606 @@
+        };
+        quote!(
+            struct #struct_name {
+                service: Arc<Box<dyn #service_name + Send + Sync>>,
```
Please keep the existing handler representation:

```rust
Arc<dyn Trait + Send + Sync>
```

instead of `Arc<Box<dyn Trait + Send + Sync>>`. The additional `Box` adds an allocation and pointer indirection and unnecessarily changes the generated public API. This applies to both the handler fields and the sync/async `create_*` functions.

Tim-Zhang commented on src/ttrpc.proto:
```diff
@@ -14,7 +14,7 @@
 syntax = "proto3";
 
-package grpc;
+package ttrpc;
```
Please keep the original `package grpc`.

Changing it to `ttrpc` also changes protobuf descriptor full names and therefore the schema/API identity. The Prost build and Rust module layout should adapt to the existing schema—for example, generate `grpc.rs` and re-export it through the desired Rust-side module—instead of modifying the protobuf package.

Tim-Zhang commented on build.rs:
```diff
@@ -20,3 +25,76 @@ fn main() {
+    // SCREAMING_SNAKE_CASE names (e.g. `OK`, `CANCELLED`) that the
+    // rustprotobuf backend keeps verbatim. Rewrite the generated variants
+    // back so both backends expose the same public `Code` API.
+    let mut content = fs::read_to_string(&ttrpc_path).expect("Failed to read ttrpc.rs");
```
Preserving names such as `Code::OK` is necessary, but rewriting generated Rust using formatting-sensitive string replacements is fragile.

Please add compatibility constants instead:

```rust
impl Code {
    pub const OK: Self = Self::Ok;
    pub const NOT_FOUND: Self = Self::NotFound;
}
```

This preserves the public API without mutating generated source text.

Tim-Zhang commented on src/lib.rs:
```diff
@@ -62,10 +62,10 @@ pub use crate::error::{get_status, Error, Result};
     #[doc(inline)]
     pub use sync::{MethodHandler, TtrpcContext};
+    #[doc(inline)]
+    pub use sync::utils::{response_to_channel, response_error_to_channel};
```
`response_error_to_channel` and `sync::utils` are implementation details and should not become new public APIs. Please keep the module/helper private and preserve the existing `#[doc(hidden)]` compatibility exports of `response_to_channel`.

Tim-Zhang commented on src/sync/utils.rs:
```diff
@@ -78,8 +102,41 @@ macro_rules! request_handler {
+#[macro_export]
+#[cfg(feature = "prost")]
+macro_rules! request_handler {
+    ($class: ident, $ctx: ident, $req: ident, $req_type: ident, $req_fn: ident) => {
```
The exported macro has different signatures depending on the protobuf backend: rust-protobuf accepts six arguments while Prost accepts five. In addition, `$req_type:ident` cannot match qualified generated types such as `super::google::protobuf::Empty`.

Please keep the existing invocation form supported and use a path-aware matcher such as `$req_type:path`, or route both backends through a shared codec abstraction. The corresponding async macro should be consistent.

Tim-Zhang commented on Cargo.toml:
```diff
@@ -60,13 +68,20 @@ windows-sys = {version = "0.48", features = [ "Win32_Foundation", "Win32_Storage
 async = ["async-trait", "async-stream", "tokio", "futures", "tokio-vsock"]
 sync = []
+prost = ["dep:prost", "dep:prost-build"]
+rustprotobuf = ["dep:protobuf"]
```
Please make `protobuf-codegen` optional as well, so Prost-only builds do not compile rust-protobuf dependencies:

```toml
[build-dependencies]
protobuf-codegen = { workspace = true, optional = true }
prost-build = { version = "0.13", optional = true }

[features]
default = ["sync", "rustprotobuf"]
rustprotobuf = ["dep:protobuf", "dep:protobuf-codegen"]
prost = ["dep:prost", "dep:prost-build"]
```

Please also use `#[cfg(feature = "rustprotobuf")]` instead of `#[cfg(not(feature = "prost"))]`, and report a clear compile error when neither backend is selected.

Tim-Zhang commented on example2/build.rs:
```diff
@@ -0,0 +1,42 @@
+use ttrpc_codegen::{AsyncMode, CodegenBuilder};
+
+fn main() {
+    let mut protos = vec![
```
Keeping a standalone Prost example crate is reasonable because the `prost` and `rustprotobuf` runtime features are mutually exclusive.

However, it should reuse the existing unchanged schemas from `../example/protocols/protos` instead of cloning the whole proto tree. The duplicated protos account for 2,170 lines and the duplicated update script for another 82 lines. Sharing them removes 2,252 of the 3,238 added example lines—about 70%—while keeping the backend-specific Rust implementation separate.

Tim-Zhang commented on example2/protocols/protos/health.proto:
```diff
@@ -0,0 +1,40 @@
+
+syntax = "proto3";
+
+package health;
```
The copied schemas change several protobuf packages: `grpc` to `health`, `agent`, and `oci`, and `ttrpc.test.streaming` to `streaming`. This changes service full names and RPC paths.

Please reuse the original schemas unchanged. Adapt the Rust imports and module glue to Prost's generated layout instead of changing protobuf packages.

Tim-Zhang commented on example2/Cargo.toml:
```diff
@@ -0,0 +1,59 @@
+[package]
+name = "ttrpc-example"
```
Please rename the `example2` directory to `example-prost`. The current name does not communicate its purpose and becomes ambiguous as examples are added.

Please update the corresponding paths and use:

```toml
name = "ttrpc-example-prost"
publish = false
```

Tim-Zhang commented on example2/client.rs:
```diff
@@ -0,0 +1,119 @@
+        "agent.online_cpu_mem()",
+        now.elapsed()
+    );
+    let show = match agent_service_client
```
Several RPC calls only format or print the `Ok`/`Err` result. This allows the process to exit successfully even when an RPC fails, so the BVT can pass without validating behavior.

Please restore meaningful success assertions and expected-error checks from the existing example so unexpected failures produce a non-zero outcome.

Tim-Zhang commented on tests/run-examples.rs:
```diff
@@ -94,17 +95,42 @@ fn wait_with_output(name: &str, cmd: Child) {
-        run_example("async-stream-server", "async-stream-client", &["--tcp"])?;
+        println!("Running examples with prost feature");
+        #[cfg(unix)]
+        run_example("async-server", "async-client", "example2", &[])?;
```
The Prost BVT runs async unary and async streaming, but the sync server/client pair is only compiled and never executed.

Please run sync unary, async unary, and async streaming where supported. The sync and async implementations have separate response and macro paths.

Tim-Zhang commented on codegen/src/svcgen.rs:
```diff
@@ -0,0 +1,606 @@
+        let service_trait = format_ident!("{}", service.name);
+        let mod_path = ttrpc_mod();
+        let method_inserts: Vec<_> = service.methods.iter().map(|method| {
+            let key = format!("/{}.{}/{}", service.package, service.name, method.proto_name);
```
Package-less services are generated with an extra dot, for example `/.Health/Check` and `.Health`. The new test only verifies that generation does not panic.

Please add the separator only when a package is present and assert the generated values `/Health/Check` and `Health`.

Tim-Zhang commented on src/sync/utils.rs:
```diff
@@ -35,6 +37,27 @@ pub fn response_to_channel(
+    tx: std::sync::mpsc::Sender<(MessageHeader, Vec<u8>)>,
+) -> Result<()> {
+    let mut buffer = Vec::new();
+    <Response as prost::Message>::encode(&res, &mut buffer).map_err(err_to_others_err!(e, ""))?;
```
The Prost sync response path bypasses the existing `MESSAGE_LENGTH_MAX` handling, including replacement of oversized responses with the corresponding error response.

Please route this through the shared codec/response path or apply the same complete encoded-message size validation as the existing backend. The async response path should follow the same rule.

Tim-Zhang commented on src/sync/client.rs:
```diff
@@ -158,6 +159,10 @@ impl Client {
     }
     pub fn request(&self, req: Request) -> Result<Response> {
+        #[cfg(feature = "prost")]
+        check_oversize(req.payload.len(), false)?;
```
Checking only `req.payload.len()` does not guarantee that the serialized `Request` fits within `MESSAGE_LENGTH_MAX`, because the envelope and protobuf length prefixes also consume space.

Please validate the complete encoded request, preferably through `Codec`, before writing the final buffer. The async client should be checked consistently.

Tim-Zhang commented on .github/workflows/bvt.yml:
```diff
@@ -27,14 +30,18 @@ jobs:
           make -C example build-examples
-        # It's important for windows to fail correctly
-        # https://github.com/actions/runner-images/issues/6668 
+          make -C example2 build-examples
```
The new `codegen` crate is excluded from the root workspace, so the current root commands do not run its unit tests.

Please explicitly add the equivalent of:

```sh
make -C codegen check
make -C codegen test
```

to CI.

Tim-Zhang reviewed (changes requested):
Thanks for the continued updates. I hope we can stay in close communication and get this PR merged within the next two weeks.

The latest revision fixes several earlier issues, and the current checks are green. I also verified locally that the codegen tests and the main library tests for both protobuf backends pass.

The main remaining items are:

1. Preserve the existing `ttrpc-codegen` fluent API and add `.prost()` for backend selection instead of requiring users to migrate to `CodegenBuilder`.
2. Avoid unnecessary generated API differences, including `Arc<Box<dyn Trait>>`, backend-dependent macro signatures, and newly exposed internal helpers.
3. Keep the original protobuf package declarations. Rust module layout differences should be handled in the build or module glue rather than by changing protocol identities.
4. Keep the standalone Prost example, but rename `example2` to `example-prost` and reuse the existing proto schemas. Removing the duplicated proto directory and update script would eliminate 2,252 of the 3,238 added example lines.
5. Make the example tests validate RPC results and execute the sync example in CI.
6. Make the codegen dependencies backend-specific, fix package-less service names, enforce size limits against complete encoded messages, and run the codegen crate tests explicitly in CI.

Acronym casing differences such as `OnlineCPUMemRequest` and `OnlineCpuMemRequest` are determined by the respective generators. I am not requesting automatic compatibility aliases for these types in this PR.

The PR also currently conflicts with the target branch and will need to be updated before merging.

Tim-Zhang commented on codegen/src/codegen.rs:
```diff
@@ -0,0 +1,271 @@
+            .map_err(|e| io::Error::other(format!("decode fd_set: {e}")))?;
+
+        for fd in fd_set.file.iter() {
+            let rs_path = self.out_dir.join(format!("{}.rs", fd.package()));
```
Multiple proto files can share the same package and therefore the same output file. Here, `health.proto`, `agent.proto`, and `oci.proto` all map to `grpc.rs`, so the generated file starts with the same header three times:

```rust
// This file is generated by ttrpc-codegen 1.0.0. Do not edit
// @generated

// This file is generated by ttrpc-codegen 1.0.0. Do not edit
// @generated

// This file is generated by ttrpc-codegen 1.0.0. Do not edit
// @generated
```

Please deduplicate `rs_path` before rewriting the file.

jokemanfire replied:
Ok done

Tim-Zhang reviewed (commented):
@jokemanfire Thanks for the updates. One last codegen issue:

Tim-Zhang reviewed (approved):
@jokemanfire Thank you so much for all the time and effort you’ve put into this PR. It has been a long journey, and I truly appreciate your patience and persistence in addressing the feedback and keeping it moving forward. This is a substantial contribution to the project and will provide a solid foundation for Prost support going forward. Great work, and thank you again!

jokemanfire commented:
> @jokemanfire Thank you so much for all the time and effort you’ve put into this PR. It has been a long journey, and I truly appreciate your patience and persistence in addressing the feedback and keeping it moving forward. This is a substantial contribution to the project and will provide a solid foundation for Prost support going forward. Great work, and thank you again!


Thank you for recognizing and guiding the work, along with the previous contributors who participated in this work.
